### PR TITLE
Implement `CustomLogPipes` for Spin Logging/IO

### PR DESCRIPTION
**Pull request by danbugs**
> That's exactly right. If you have your WasiFiles already, you can call `ModuleIoRedirects::new` directly, and pass that to `prepare_component`. (At least, you'll be able to if that PR is accepted.)
> 
> WAGI does this in a very simple form because it needs different behaviour for stdin, and a slightly different interaction with the buffer for stdout. It's really valuable to know that this is more broadly useful as a public customisation point though!

To provide an update on this from #438 (and maybe close that issue soon), it turns out that I couldn't really make the current architecture use our own pipes — As previously implied, the meat and potatoes of logging is done in `execute`:
https://github.com/fermyon/spin/blob/c3ff7d90fe115daa3fa85ce3e626c738e7baef76/crates/http/src/spin.rs#L19-L59

It's there that `capture_io_to_memory`, `prepare_component` (that receives redirects), and, subsequently, `redirect_to_mem_buffer` are called.

Considering our builder is set w/ `with_engine()`, as per:
https://github.com/fermyon/spin/blob/1fbfb23b4e96a580d6d4f4b6d617e916974493f1/crates/engine/src/lib.rs#L103-L119

I tried emulating the work done in `execute` in the `build` stage similar to how the `store` does it:
https://github.com/fermyon/spin/blob/0f51ad0b86b798ae18cb4f211f0c3fcfbbe9a156/crates/engine/src/lib.rs#L304-L316

... but, nothing came of it as `execute` would eventually run and override/never access my `RuntimeContext` WASI context data.

All and all, this PR allows users to set `CustomLogPipes` when creating their `ExecutionContextConfiguration` — these `CustomLogPipes` are later accessed by `execute` and passed in to `capture_io_to_memory` (which has a side-effect on `prepare_component`) providing a good and needed override of Spin's default settings.

### CODE REVIEW

**Comment by Mossaka** [comment]
@lann @itowlson @radu-matei for review~

File: crates/engine/src/io.rs [hunk 1]
```diff
@@ -38,6 +40,63 @@ pub trait OutputBuffers {
...
+                    .unwrap();
+        Self(f, self.1.clone())
+    }
+}
+
+/// CustomIoPipes that can be passed to `ExecutionContextConfiguration`
+/// to direct out and err
+#[derive(Clone, Debug)]
+pub struct CustomLogPipes {
+    pub(crate) stdout_pipe: PipeFile,
```
Comment (Mossaka) [inline crates/engine/src/io.rs:83]: what's the reason for leaving stdin_pipe behind?

Comment (danbugs) [reply crates/engine/src/io.rs]: I've just added it!

**Comment by danbugs** [comment]
_Note: I had to do a force push to fix commits that weren't signed-off~ I was under the impression that having a `git config --global` set w/ `user.email` and `user.name` made it so that `-s` in `git commit` wasn't necessary 🤷_

**Comment by danbugs** [comment]
⚠️I think my rebases made this PR a bit messy, so I remade it here at #482 for ease of review~

**Pull request closed without merging**